coding/coding.go

```diff
@@ -1,75 +1,112 @@
-package ipfsld
+package coding
 
 import (
-	mc "github.com/jbenet/go-multicodec"
-	mccbor "github.com/jbenet/go-multicodec/cbor"
-	mcmux "github.com/jbenet/go-multicodec/mux"
+	"bytes"
+	"fmt"
+	"io"
 
-	ipld "github.com/ipfs/go-ipld"
+	cbor "github.com/ipfs/go-ipld/coding/cbor"
+	json "github.com/ipfs/go-ipld/coding/json"
 	pb "github.com/ipfs/go-ipld/coding/pb"
+
+	memory "github.com/ipfs/go-ipld/memory"
+	stream "github.com/ipfs/go-ipld/stream"
+	mc "github.com/jbenet/go-multicodec"
+)
+
+var Header []byte
+
+const (
+	HeaderPath = "/mdagv1"
 )
 
-// defaultCodec is the default applied if user does not specify a codec.
-// Most new objects will never specify a codec. We track the codecs with
-// the object so that multiple people using the same object will continue
-// to marshal using the same codec. the only reason this is important is
-// that the hashes must be the same.
-var defaultCodec string
+type Codec int
 
-var muxCodec *mcmux.Multicodec
+const (
+	NoCodec       Codec = 0
+	CodecProtobuf Codec = iota
+	CodecCBOR
+	CodecJSON
+	CodecCBORNoTags
+)
+
+var StreamCodecs map[string]func(io.ReadSeeker) (stream.NodeReader, error)
```

Comment:
There are two things here:

- extensibility is generally better in terms of objects in a OOP language (I would argue with that, especially with a language like Go which allow nice function expressions, but never mind)
- readability. In this case, a single literal expression can be used to declare all codecs. That's really nice and makes it easier to read. Also, it is very easy to know which code is called in a function call. With too much OOP interfaces, it becomes very difficult.

About the multicodec package, I tried first to use it, but it isn't suited to work with streams. See [my last message](#issuecomment-191811920)